src/context.rs

```diff
@@ -235,8 +235,31 @@ impl<'a> Context<'a> {
         disabled == Some(true)
     }
 
+    /// Returns true when a negated environment variable is defined in `env_vars` and is present
+    fn has_negated_env_var(&self, env_vars: &'a [&'a str]) -> bool {
+        env_vars
+            .iter()
+            .any(|env_var| env_var.starts_with('!') && self.get_env(&env_var[1..]).is_some())
+    }
+
+    /// Returns true if 'detect_env_vars' is empty,
+    /// or if at least one environment variable is set and no negated environment variable is set
+    /// or if there are only negated environment variable provided (which are not set)
     pub fn detect_env_vars(&'a self, env_vars: &'a [&'a str]) -> bool {
-        env_vars.is_empty() || (env_vars.iter().any(|e| self.get_env(e).is_some()))
+        if env_vars.is_empty() {
+            return true;
+        }
+
+        if self.has_negated_env_var(env_vars) {
+            return false;
+        }
+
+        // Returns true if all environment variables start with a "!",
+        // or if at least one environment variable is set
+        env_vars.iter().all(|env_var| env_var.starts_with('!'))
+            || env_vars
+                .iter()
+                .any(|env_var| self.get_env(env_var).is_some())
```

Comment:
What do you think about this?

```suggestion
        let mut iter = env_vars
            .iter()
            .filter(|env_var| !env_var.starts_with('!'))
            .peekable();

        iter.peek().is_none() || iter.any(|env_var| self.get_env(env_var).is_some())
```

Reply:
Faster and more readable as well. I just need to rewrite the comments.